wiring/src/spark_wiring_cloud_event.cpp

```diff
@@ -0,0 +1,778 @@
+    int r = decodeFromCBOR(v, *this);
+    if (r < 0) {
+        LOG(ERROR, "decodeFromCBOR() failed: %d", r);
+        return Variant();
+    }
+    return v;
+}
+
+CloudEvent& CloudEvent::maxDataSizeInRam(size_t size) {
+    if (!isWritable() || d_->payload) {
+        return *this;
+    }
+    d_->maxHeapSize = std::min(size, MAX_DATA_SIZE);
+    return *this;
+}
+
+size_t CloudEvent::maxDataSizeInRam() const {
+    if (!d_) {
+        return DEFAULT_MAX_PAYLOAD_HEAP_SIZE;
+    }
+    return d_->maxHeapSize;
+}
+
+CloudEvent& CloudEvent::onStatusChange(std::function<OnStatusChange> callback) {
+    if (!isWritable()) {
+        return *this;
+    }
+    d_->onStatusChange = std::move(callback);
+    return *this;
+}
+
+CloudEvent::Status CloudEvent::status() const {
+    if (!d_) {
+        return Status::INVALID;
+    }
+    return d_->status;
+}
+
+int CloudEvent::error() const {
+    if (!d_) {
+        return Error::NO_MEMORY;
+    }
+    return d_->error;
+}
+
+void CloudEvent::cancel() {
+    if (!d_ || d_->status != Status::SENDING) {
+        return;
+    }
+    coap_cancel_request(d_->requestId, nullptr /* reserved */);
+    d_->requestId = COAP_INVALID_REQUEST_ID;
+    RateLimiter::instance().give(this->size());
+    // TODO: For now, transition to an invalid state as an event in a failed state can be sent again
+    // and that would create a race condition between cancellation and normal completion of the event
+    // in sendComplete()
+    setInvalid(Error::CANCELLED);
+}
+
+void CloudEvent::reset() {
+    d_ = makeRefCountPtr<Data>();
+}
+
+int CloudEvent::available() {
+    if (!isReadable() || !d_->payload) {
+        return 0;
+    }
+    int r = coap_get_payload_size(d_->payload.get(), nullptr /* reserved */);
+    if (r < 0) {
+        LOG(ERROR, "coap_get_payload_size() failed: %d", r);
+        return 0;
+    }
+    return r - d_->pos;
+}
+
+int CloudEvent::read(char* data, size_t size) {
+    int r = peek(data, size);
+    if (r < 0) {
+        return r;
+    }
+    d_->pos += r;
+    return r;
+}
+
+int CloudEvent::peek(char* data, size_t size) {
+    if (!isReadable()) {
+        return error();
+    }
+    if (!d_->payload) {
+        if (size > 0) {
+            return Error::END_OF_STREAM;
+        }
+        return 0;
+    }
+    int r = coap_read_payload(d_->payload.get(), data, size, d_->pos, nullptr /* reserved */);
+    if (r < 0 && r != Error::END_OF_STREAM) {
+        LOG(ERROR, "coap_read_payload() failed: %d", r);
+    }
+    return r;
+}
+
+int CloudEvent::write(const char* data, size_t size) {
+    if (!isWritable()) {
+        int err = error();
+        if (err < 0) {
+            return err;
+        }
+        return Error::INVALID_STATE;
+    }
+    auto payload = getValidPayload();
+    if (!payload) {
+        return error();
+    }
+    int r = coap_write_payload(payload, data, size, d_->pos, nullptr /* reserved */);
+    if (r < 0) {
+        if (r == Error::COAP_TOO_LARGE_PAYLOAD) {
```

Comment:
Is there a way for an application to know how much room is left on the event (or what's the maximum total size of a large event) before calling write and getting COAP_TOO_LARGE_PAYLOAD?

Reply:
The maximum supported size is exposed as `CloudEvent::MAX_DATA_SIZE`. The app can check how much space is left like this:
```cpp
if (event.pos() + 100 <= CloudEvent::MAX_DATA_SIZE) {
    event.write(data, 100);
}
```